integration/e2e/limits_test.go

```diff
@@ -90,6 +98,104 @@ func TestLimits(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestOTLPLimits(t *testing.T) {
+	s, err := e2e.NewScenario("tempo_e2e")
+	require.NoError(t, err)
+	defer s.Close()
+
+	require.NoError(t, util.CopyFileToSharedDir(s, configLimits, "config.yaml"))
+	tempo := util.NewTempoAllInOne()
+	require.NoError(t, s.StartAndWaitReady(tempo))
+
+	protoSpans := test.MakeProtoSpans(100)
+
+	// gRPC
+	grpcClient := otlptracegrpc.NewClient(
+		otlptracegrpc.WithEndpoint(tempo.Endpoint(4317)),
+		otlptracegrpc.WithInsecure(),
+		otlptracegrpc.WithRetry(otlptracegrpc.RetryConfig{Enabled: false}),
+	)
+	require.NoError(t, grpcClient.Start(context.Background()))
+
+	grpcErr := grpcClient.UploadTraces(context.Background(), protoSpans)
+	assert.Error(t, grpcErr)
+	require.Equal(t, codes.ResourceExhausted, status.Code(grpcErr))
+
+	// HTTP
+	httpClient := otlptracehttp.NewClient(
+		otlptracehttp.WithEndpoint(tempo.Endpoint(4318)),
+		otlptracehttp.WithInsecure(),
+		otlptracehttp.WithRetry(otlptracehttp.RetryConfig{Enabled: false}),
+	)
+	require.NoError(t, httpClient.Start(context.Background()))
+
+	httpErr := httpClient.UploadTraces(context.Background(), protoSpans)
+	assert.Error(t, httpErr)
+	require.Contains(t, httpErr.Error(), "retry-able request failure")
+}
+
+func TestOTLPLimitsVanillaClient(t *testing.T) {
+	s, err := e2e.NewScenario("tempo_e2e")
+	require.NoError(t, err)
+	defer s.Close()
+
+	require.NoError(t, util.CopyFileToSharedDir(s, configLimits, "config.yaml"))
+	tempo := util.NewTempoAllInOne()
+	require.NoError(t, s.StartAndWaitReady(tempo))
+
+	trace := test.MakeTrace(10, []byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10})
+
+	testCases := []struct {
+		name    string
+		payload func() []byte
+		headers map[string]string
+	}{
+		// TODO There is an issue when sending the payload in json format. The server returns a 200 instead of a 429.
+		// {
+		// 	"JSON format",
+		// 	func() []byte {
+		// 		b := &bytes.Buffer{}
+		// 		err := (&jsonpb.Marshaler{}).Marshal(b, trace)
+		// 		require.NoError(t, err)
+		// 		return b.Bytes()
+		// 	},
+		// 	map[string]string{
+		// 		"Content-Type": "application/json",
+		// 	},
+		// },
+		{
+			"Proto format",
+			func() []byte {
+				b, err := trace.Marshal()
+				require.NoError(t, err)
+				return b
+			},
+			map[string]string{
+				"Content-Type": "application/x-protobuf",
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			req, err := http.NewRequest(http.MethodPost, "http://"+tempo.Endpoint(4318)+"/v1/traces", bytes.NewReader(tc.payload()))
+			require.NoError(t, err)
+			for k, v := range tc.headers {
+				req.Header.Set(k, v)
+			}
+
+			resp, err := http.DefaultClient.Do(req)
+			require.NoError(t, err)
+			defer func() { _ = resp.Body.Close() }()
+			bodyBytes, err := io.ReadAll(resp.Body)
+			require.NoError(t, err)
+			fmt.Println(string(bodyBytes))
+
+			assert.Equal(t, http.StatusTooManyRequests, resp.StatusCode)
```

Comment:
Check my understanding, but we are sending multiple traces in one reqeust, and the config specifies a limit of one trace per user.  Should this fail with "TooManyRequests" since we are only making one request?

Reply:
The config limits fixture file used in this tests specifies the following https://github.com/mapno/tempo/blob/70bf4e253df931f68e747b267c496adaa8e77f6a/integration/e2e/config-limits.yaml#L20

```
overrides:
  defaults:
    ingestion:
      max_traces_per_user: 1
      rate_limit_bytes: 500
      burst_size_bytes: 500
    global:
      max_bytes_per_trace: 130
```

The test is batching up 10 traces
https://github.com/mapno/tempo/blob/5faf0f091584b25e43f57fdcb27e5f2da7cc342a/integration/e2e/limits_test.go#L146

So the quota is being hit and the status code returned is 429 which is known as Resource Exhausted but in go its also TooManyRequests https://github.com/golang/go/blob/master/src/net/http/status.go#L62